src/dmd/cppmangle.d

```diff
@@ -1373,4 +1453,400 @@ extern(C++):
+        }
+        buf.writeByte('I');
+        // Need to check with the declaration what we have on our hands
+        auto analyzed_ti = this.context.res.asType.toDsymbol(null).isInstantiated();
+        auto prev = this.context;
+        scope (exit) this.context.reset(prev);
+        foreach (idx, RootObject o; *t.tiargs)
+        {
+            this.context.res = (*analyzed_ti.tiargs)[idx];
+            o.visitObject(this);
+        }
+        if (analyzed_ti.tiargs.dim > t.tiargs.dim)
+        {
+            // If the resolved AST has more args than the parse one,
+            // we have default arguments
+            auto oparams = (cast(TemplateDeclaration)analyzed_ti.tempdecl).origParameters;
+            foreach (idx, arg; (*oparams)[t.tiargs.dim .. $])
+            {
+                this.context.res = (*analyzed_ti.tiargs)[idx + t.tiargs.dim];
+
+                if (auto ttp = arg.isTemplateTypeParameter())
+                    ttp.defaultType.accept(this);
+                else if (auto tvp = arg.isTemplateValueParameter())
+                    tvp.defaultValue.accept(this);
+                else if (auto tvp = arg.isTemplateThisParameter())
+                    tvp.defaultType.accept(this);
+                else if (auto tvp = arg.isTemplateAliasParameter())
+                    tvp.defaultAlias.visitObject(this);
+                else
+                    assert(0, arg.toString());
+            }
+        }
+        buf.writeByte('E');
+        if (needclosing)
+            buf.writeByte('E');
+    }
+
+    /// Ditto
+    override void visit(IntegerExp t)
+    {
+        this.buf.writeByte('L');
+        t.type.accept(this);
+        this.buf.print(t.getInteger());
+        this.buf.writeByte('E');
+    }
+
+    override void visit(Nspace t)
+    {
+        if (auto p = getQualifier(t))
+            p.accept(this);
+
+        if (isStd(t))
+            buf.writestring("St");
+        else
+        {
+            this.writeIdentifier(t.ident);
+            this.append(t);
+        }
+    }
+
+    override void visit(Type t)
+    {
+        error(t);
+    }
+
+    void visit(Tuple t)
+    {
+        assert(0);
+    }
+
+    /**
+     * Helper function to go through the `TemplateParameter`s
+     * and perform a substitution, if possible.
+     *
+     * Returns:
+     *   The `TemplateParameter` that was used, or `null`.
+     */
+    private bool writeTemplateSubstitution (
+        const ref Identifier ident, TemplateParameters* params, Type type)
+    {
+        foreach (idx, param; *params)
+        {
+            if (param.ident == ident)
+            {
+                if (type)
+                    CV_qualifiers(type);
+                if (this.substitute(param))
+                    return true;
+                this.append(param);
+
+                // Since we didn't run semantic yet, all we get are
+                // `TypeIdentifier`, but we need to check if it's an expression
+                // or whatnot
```

Comment:
This (and maybe other places, I haven't looked too deep), could be rephrased to use a little less `we`.

For instance (feel free to correct as you see fit):
```
Since semantic hasn't run yet, all that is handled here are `TypeIdentifier`,
but we still need to check if it's an expression or not.
```